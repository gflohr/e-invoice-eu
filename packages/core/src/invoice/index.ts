/* istanbul ignore file */
export type { Invoice } from './invoice.interface';
export * from './invoice.schema';
export * from './invoice.service';
