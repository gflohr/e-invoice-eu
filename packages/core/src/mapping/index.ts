/* istanbul ignore file */

export type { Mapping } from './mapping.interface';
export * from './mapping.regex';
export * from './mapping.schema';
export * from './mapping.service';
