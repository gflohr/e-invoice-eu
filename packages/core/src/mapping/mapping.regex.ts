// A sheet Ref is anything inside single quotes or a string without colons and dots.
const quotedSheetRef = "'[^']+'";
const unquotedSheetRef = '[^:.]+';
const sheetRef = `(${unquotedSheetRef}|${quotedSheetRef})?`; // Optional and captured

// A section Ref is a colon followed by anything but a colon or dot.
const sectionRef = '([^.:]+)?'; // Optional section ref

// Column and row names
const columnName = '[A-Z]+';
const rowName = '[1-9][0-9]*';
const cellName = `(${columnName}${rowName})`; // Capture group for cell reference

// Cell reference
const cellRef = `=(?:${sheetRef}(?::${sectionRef})?\\.)?${cellName}`;

// A leading single-quote is ignored but forces textual interpretation.
const quotedLiteral = "'[^']+'";
const unquotedLiteral = '[^=].*';
const literal = `(${quotedLiteral}|${unquotedLiteral})`; // Capture group for literal

/**
 * This regular expression matches a value reference, i.e. the value of a node
 * in the mapping object. They are always strings!
 *
 * The captures are:
 *
 * 0: the entire reference 1: a possibly single-quoted sheet name or undefined
 * 2: a section name without the leading colon or undefined 3: a cell name
 * without the leading dot or undefined 4: a literal value, possibly with a
 * leading single-quote, or undefined
 */
export const mappingValuePattern = `^${cellRef}|${literal}$`;
export const mappingValueRe = new RegExp(mappingValuePattern);

/**
 * This regular expression matches a section Reference, i.e. the value of the
 * property `section` for arrays. It is an optional, possibly quoted sheet
 * reference followed by a colon and the name of a section. A lone section must
 * start with a colon. The catptures are:
 *
 * 0: the entire reference 1: a possibly quoted sheet reference or undefined 2:
 * a section reference without the colon
 */
export const sectionReferencePattern = `(?:${sheetRef}(?::${sectionRef}))`;
export const sectionReferenceRe = new RegExp(sectionReferencePattern);
