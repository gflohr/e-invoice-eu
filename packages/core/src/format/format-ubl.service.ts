import { Mapping } from '@e-invoice-eu/core';
import {
	ADDITIONALSUPPORTINGDOCUMENTS,
	AttachedDocumentMimeCode,
} from '@e-invoice-eu/core/invoice/invoice.interface';

import { FormatXMLService } from './format-xml.service';
import { EInvoiceFormat } from './format.e-invoice-format.interface';
import { Invoice } from '../invoice/invoice.interface';
import { invoiceSchema } from '../invoice/invoice.schema';
import { InvoiceServiceOptions } from '../invoice/invoice.service';
import { sortBySchema } from '../utils/sort-by-schema';

export class FormatUBLService
	extends FormatXMLService
	implements EInvoiceFormat
{
	get customizationID(): string {
		return 'urn:cen.eu:en16931:2017';
	}

	get profileID(): string {
		return 'urn:fdc:peppol.eu:2017:poacc:billing:01:1.0';
	}

	get syntax(): 'UBL' | 'CII' {
		return 'UBL';
	}

	fillMappingDefaults(mapping: Mapping) {
		if (!('cbc:CustomizationID' in mapping['ubl:Invoice'])) {
			mapping['ubl:Invoice']['cbc:CustomizationID'] = this.customizationID;
		}

		if (!('cbc:ProfileID' in mapping['ubl:Invoice'])) {
			mapping['ubl:Invoice']['cbc:ProfileID'] = this.profileID;
		}

		const orderReference = mapping['ubl:Invoice']['cac:OrderReference'];
		if (
			orderReference &&
			'cbc:SalesOrderID' in orderReference &&
			!('cbc:ID' in orderReference)
		) {
			mapping['ubl:Invoice']['cac:OrderReference']!['cbc:ID']! = 'NA';
		}
	}

	fillInvoiceDefaults(invoice: Invoice) {
		if (!('cbc:customizationID' in invoice['ubl:Invoice'])) {
			invoice['ubl:Invoice']['cbc:CustomizationID'] = this.customizationID;
		}

		if (!('cbc:profileID' in invoice['ubl:Invoice'])) {
			invoice['ubl:Invoice']['cbc:ProfileID'] = this.profileID;
		}

		const orderReference = invoice['ubl:Invoice']['cac:OrderReference'];
		if (
			orderReference &&
			'cbc:SalesOrderID' in orderReference &&
			!('cbc:ID' in orderReference)
		) {
			orderReference['cbc:ID'] = 'NA';
		}
	}

	async generate(
		invoice: Invoice,
		options: InvoiceServiceOptions,
	): Promise<string | Uint8Array> {
		await this.embedPDF(invoice, options);
		this.embedAttachments(invoice, options);

		invoice = sortBySchema(invoice, invoiceSchema);

		const cnCodes = invoiceSchema['$defs']!.codeLists['UNCL1001-cn'].enum;
		const code = invoice['ubl:Invoice']['cbc:InvoiceTypeCode'];
		if (cnCodes.includes(code) && code !== '384') {
			const creditNoteObject = {
				CreditNote: {
					'@xmlns': 'urn:oasis:names:specification:ubl:schema:xsd:CreditNote-2',
					'@xmlns:cac':
						'urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2',
					'@xmlns:cbc':
						'urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2',
					...invoice['ubl:Invoice'],
				},
			};

			(invoice as any)['ubl:CreditNote'] = invoice['ubl:Invoice'];
			delete (invoice as any)['ubl:Invoice'];

			function replaceInvoiceWithCreditNote(obj: object): object {
				// Recursively clone and transform the object
				if (Array.isArray(obj)) {
					return obj.map(replaceInvoiceWithCreditNote);
				} else if (obj !== null && typeof obj === 'object') {
					const newObj: object = {};
					for (const [key, value] of Object.entries(obj)) {
						let newKey = key;
						if (
							key.includes(':Invoice') &&
							!key.includes(':InvoicePeriod') &&
							!key.includes(':InvoiceDocumentReference')
						) {
							if (key.includes(':InvoicedQuantity')) {
								newKey = key.replace(':Invoiced', ':Credited');
							} else {
								newKey = key.replace(':Invoice', ':CreditNote');
							}
						}
						newObj[newKey] = replaceInvoiceWithCreditNote(value);
					}
					return newObj;
				}
				return obj;
			}

			return this.renderXML(replaceInvoiceWithCreditNote(creditNoteObject));
		} else {
			const invoiceObject = {
				Invoice: {
					'@xmlns': 'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2',
					'@xmlns:cac':
						'urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2',
					'@xmlns:cbc':
						'urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2',
					...invoice['ubl:Invoice'],
				},
			};

			return this.renderXML(invoiceObject);
		}
	}

	private async embedPDF(invoice: Invoice, options: InvoiceServiceOptions) {
		if (!options.embedPDF) return;

		const pdf = await this.getInvoicePdf(options);
		const mimeType = 'application/pdf';
		let filename: string;
		if (options.pdf) {
			if (typeof options.pdf.filename === 'undefined') {
				throw new Error('A PDF filename is required!');
			}
			filename = options.pdf.filename;
		} else if (options.spreadsheet) {
			// Avoid using path.parse() here because it is not available in
			// the browser.
			const normalized = options.spreadsheet.filename.replace(/\\/g, '/');
			const lastSlashIndex = normalized.lastIndexOf('/');
			const lastDotIndex = normalized.lastIndexOf('.');
			const name =
				lastDotIndex > lastSlashIndex
					? normalized.substring(lastSlashIndex + 1, lastDotIndex)
					: normalized;

			filename = name + '.pdf';
		} // No else because this is already checked in getInvoicePdf().

		const ref: ADDITIONALSUPPORTINGDOCUMENTS = {
			'cbc:ID': options.pdf?.id ?? invoice['ubl:Invoice']['cbc:ID'],
			'cac:Attachment': {
				'cbc:EmbeddedDocumentBinaryObject': this.uint8ArrayToBase64(pdf),
				'cbc:EmbeddedDocumentBinaryObject@filename': filename!,
				'cbc:EmbeddedDocumentBinaryObject@mimeCode': mimeType,
			},
		};

		if (typeof options.pdf?.description !== 'undefined') {
			ref['cbc:DocumentDescription'] = options.pdf.description;
		}
		invoice['ubl:Invoice']['cac:AdditionalDocumentReference'] ??= [];
		invoice['ubl:Invoice']['cac:AdditionalDocumentReference'].push(ref);
	}

	private embedAttachments(invoice: Invoice, options: InvoiceServiceOptions) {
		const validMimeCodes: Set<AttachedDocumentMimeCode> = new Set([
			'text/csv',
			'application/pdf',
			'image/png',
			'image/jpeg',
			'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
			'application/vnd.oasis.opendocument.spreadsheet',
		]);

		for (const attachment of options.attachments || []) {
			// FIXME! Guess MIME type from filename if missing.
			const mimeType: AttachedDocumentMimeCode =
				attachment.mimetype as AttachedDocumentMimeCode;
			if (!validMimeCodes.has(mimeType)) {
				throw new Error(
					`The attachment MIME type '${mimeType}' is not allowed!`,
				);
			}
			const filename = attachment.filename;
			if (typeof filename === 'undefined') {
				throw new Error('The attachment filename is required!');
			}

			if (typeof attachment.buffer === 'undefined') {
				throw new Error('The attachment buffer is required!');
			}

			const id = attachment.id ?? filename;

			const ref: ADDITIONALSUPPORTINGDOCUMENTS = {
				'cbc:ID': id,
				'cac:Attachment': {
					'cbc:EmbeddedDocumentBinaryObject': this.uint8ArrayToBase64(
						attachment.buffer,
					),
					'cbc:EmbeddedDocumentBinaryObject@filename': filename,
					'cbc:EmbeddedDocumentBinaryObject@mimeCode': mimeType,
				},
			};

			if (typeof attachment.description !== 'undefined') {
				ref['cbc:DocumentDescription'] = attachment.description;
			}
			invoice['ubl:Invoice']['cac:AdditionalDocumentReference'] ??= [];
			invoice['ubl:Invoice']['cac:AdditionalDocumentReference'].push(ref);
		}
	}
}
