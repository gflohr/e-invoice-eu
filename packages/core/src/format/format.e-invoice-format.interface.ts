import { EInvoiceMIMEType } from './format.factory.service';
import { Invoice } from '../invoice/invoice.interface';
import { InvoiceServiceOptions } from '../invoice/invoice.service';
import { Mapping } from '../mapping/mapping.interface';

/** Interface representing an e-invoice format generation service. */
export interface EInvoiceFormat {
	/**
	 * The default customization ID of the format.
	 *
	 * This ID specifies a particular customization of the invoice format, which
	 * may define additional constraints or extensions to the standard.
	 */
	get customizationID(): string;

	/**
	 * The default profile ID of the format.
	 *
	 * The profile ID identifies a specific usage profile for the e-invoice, often
	 * used to distinguish between different industry-specific implementations.
	 */
	get profileID(): string;

	/**
	 * The MIME type associated with this invoice format.
	 *
	 * This indicates the expected media type when generating or processing
	 * invoices in this format.
	 */
	get mimeType(): EInvoiceMIMEType;

	/**
	 * The syntax used for this e-invoice format.
	 *
	 * This is either 'UBL' (Universal Business Language) or 'CII' (Cross Industry
	 * Invoice), which are both common standards for electronic invoicing.
	 */
	get syntax(): 'UBL' | 'CII';

	/**
	 * Populates a mapping with default and computable values.
	 *
	 * This method ensures that all required fields in the mapping have reasonable
	 * defaults before further processing.
	 *
	 * @param mapping - The mapping definition to be updated.
	 */
	fillMappingDefaults(mapping: Mapping): void;

	/**
	 * Populates an invoice with default and computable values.
	 *
	 * This method fills in any missing fields in the invoice with default values
	 * to ensure it meets the required specifications.
	 *
	 * @param invoice - The invoice to be updated.
	 */
	fillInvoiceDefaults(invoice: Invoice): void;

	/**
	 * Generates an invoice in the specified format.
	 *
	 * This function converts the given invoice data into a serialized format
	 * (such as XML or JSON) that conforms to the selected e-invoice standard.
	 *
	 * @param invoice - The invoice data structure containing all required
	 *   information.
	 * @param options - Additional options influencing invoice generation.
	 * @returns A promise resolving to the generated invoice as a string or a
	 *   Uint8Array, depending on the format.
	 */
	generate(
		invoice: Invoice,
		options: InvoiceServiceOptions,
	): Promise<string | Uint8Array>;
}
