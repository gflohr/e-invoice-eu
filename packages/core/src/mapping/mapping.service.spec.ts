import * as XLSX from '@e965/xlsx';
import { JSONSchemaType } from 'ajv';

import { MappingService } from './mapping.service';
import { EInvoiceFormat } from '../format';
import { Mapping, MappingMetaInformation } from './mapping.interface';
import { FormatFactoryService } from '../format/format.factory.service';
import { Invoice } from '../invoice';
import { Logger } from '../logger.interface';
import { ValidationService } from '../validation';

jest.mock('fs/promises');

// Used for testing `transform()`.
const mapping = {
	meta: {
		sectionColumn: {
			Unused: 'L',
			Invoice: 'K',
		},
	},
	'ubl:Invoice': {
		'cbc:CustomizationID': '=B1',
		'cbc:ID': '1234567890',
		'cbc:IssueDate': '=A1',
		'cbc:DueDate': "='Invoice'.A2",
		'cbc:Note': "'= not equals eagles",
		'cac:OrderReference': {
			'cbc:ID': '=A3',
		},
		'cac:PaymentMeans': {
			// No section.
			'cbc:PaymentMeansCode': '30',
		},
		'cac:TaxTotal': {
			'cbc:TaxAmount@currencyID': 'EUR',
		},
		'cac:LegalMonetaryTotal': {
			'cbc:LineExtensionAmount': '=:SubTotal.F1',
		},
		'cac:InvoiceLine': {
			section: ':Line',
			'cbc:ID': '=:Line.A1',
			'cac:AllowanceCharge': {
				section: ':ACLine',
				// This is relative to the allowanche/charge section.
				'cbc:Amount': '=:ACLine.B1',
				// And this is relative to the invoice line section.
				'cbc:Amount@currencyID': '=:Line.B1',
			},
		},
	},
} as unknown as Mapping;

const workbook = {
	SheetNames: ['Invoice'],
	Sheets: {
		Invoice: {
			A1: {
				t: 's',
				v: '2024-10-21',
			},
			A2: {
				t: 'd',
				v: new Date('2024-10-28T12:34:56Z'),
			},
			A3: {
				t: 's',
				v: '0815',
			},
			B1: {
				t: 's',
				v: 'urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0',
			},
			A20: {
				t: 's',
				v: '1',
			},
			A22: {
				t: 's',
				v: '2',
			},
			A23: {
				t: 's',
				v: '3',
			},
			B20: {
				t: 's',
				v: 'EUR',
			},
			B21: {
				t: 's',
				v: '23.04',
			},
			B22: {
				t: 's',
				v: 'EUR',
			},
			B23: {
				t: 's',
				v: 'EUR',
			},
			B24: {
				t: 's',
				v: '13.03',
			},
			B25: {
				t: 's',
				v: '42.00',
			},
			F27: {
				t: 's',
				v: '12345.67',
			},
			K20: {
				t: 's',
				v: 'Line',
			},
			K21: {
				t: 's',
				v: 'ACLine',
			},
			K22: {
				t: 's',
				v: 'Line',
			},
			K23: {
				t: 's',
				v: 'Line',
			},
			K24: {
				t: 's',
				v: 'ACLine',
			},
			K25: {
				t: 's',
				v: 'ACLine',
			},
			K27: {
				t: 's',
				v: 'SubTotal',
			},
			'!ref': 'A1:Z999',
		},
	},
} as XLSX.WorkBook;

const defaultMappingContext = {
	meta: {} as MappingMetaInformation,
	workbook,
	schemaPath: ['ubl:Invoice', 'cbc:ID'],
	sectionRanges: {},
	arrayPath: [],
	rowRange: [1, Infinity] as [number, number],
};

describe('MappingService', () => {
	let service: MappingService;
	const logger = {
		error: jest.fn(),
	};

	beforeEach(async () => {
		service = new MappingService(logger as unknown as Logger);
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('should be defined', () => {
		expect(service).toBeDefined();
	});

	describe('v1', () => {
		describe('should parse mappings and map data', () => {
			it('should throw an exception if a non-existing sheet is referenced', () => {
				const wb: XLSX.WorkBook = {
					Sheets: {},
				} as XLSX.WorkBook;

				try {
					service['resolveValue']('=Inwoice.A1', {} as JSONSchemaType<any>, {
						...defaultMappingContext,
						schemaPath: ['properties', 'ubl:Invoice', 'properties', 'cbc:ID'],
						workbook: wb,
					});
					throw new Error('no exception thrown');
				} catch (e) {
					expect(e).toBeDefined();
					expect(e.validation).toBeTruthy();
					expect(e.ajv).toBeTruthy();
					expect(Array.isArray(e.errors)).toBeTruthy();
					expect(e.errors.length).toBe(1);

					const error = e.errors[0];
					expect(error.instancePath).toBe('/ubl:Invoice/cbc:ID');
					expect(error.schemaPath).toBe(
						'#/properties/ubl%3AInvoice/properties/cbc%3AID',
					);
					expect(error.keyword).toBe('type');
					expect(error.params).toEqual({ type: 'string' });
					expect(error.message).toBe(
						"reference '=Inwoice.A1' resolves to null: no such sheet 'Inwoice'",
					);
				}
			});

			it('should return an empty string if a non-existing cell is referenced', () => {
				const wb: XLSX.WorkBook = {
					Sheets: {
						Invoice: {},
					},
					SheetNames: ['Invoice'],
				} as XLSX.WorkBook;
				const value = service['resolveValue'](
					'=ET742',
					{} as JSONSchemaType<any>,
					{
						...defaultMappingContext,
						schemaPath: ['properties', 'ubl:Invoice', 'properties', 'cbc:ID'],
						workbook: wb,
					},
				);
				expect(value).toBe('');
			});

			it('should return an empty string for configured empty values', () => {
				const wb: XLSX.WorkBook = {
					Sheets: {
						Invoice: {
							A1: {
								t: 's',
								v: '[:empty:]',
							},
							A2: {
								t: 's',
								v: 'not empty',
							},
						},
					},
					SheetNames: ['Invoice'],
				} as XLSX.WorkBook;
				const empty = service['resolveValue'](
					'=A1',
					{} as JSONSchemaType<any>,
					{
						...defaultMappingContext,
						meta: {
							sectionColumn: {},
							empty: ['[:empty:]'],
						},
						schemaPath: ['properties', 'ubl:Invoice', 'properties', 'cbc:ID'],
						workbook: wb,
					},
				);
				expect(empty).toBe('');

				const notEmpty = service['resolveValue'](
					'=A2',
					{} as JSONSchemaType<any>,
					{
						...defaultMappingContext,
						schemaPath: ['properties', 'ubl:Invoice', 'properties', 'cbc:ID'],
						workbook: wb,
					},
				);
				expect(notEmpty).toBe('not empty');
			});

			it('should throw an exception if a non-existing section is referenced', async () => {
				const localMapping = structuredClone(mapping);
				localMapping['ubl:Invoice']['cac:InvoiceLine']['cbc:ID'] = '=:Lines.A1';
				const mockValidateMapping = jest
					.spyOn(service as any, 'validateMapping')
					.mockReturnValue(localMapping);
				const buf: Uint8Array = [] as unknown as Uint8Array;

				jest.spyOn(XLSX, 'read').mockReturnValueOnce(workbook);

				try {
					service.transform(buf, 'UBL', {} as Mapping);
					throw new Error('no exception thrown');
				} catch (e) {
					expect(e).toBeDefined();
					expect(e.validation).toBeTruthy();
					expect(e.ajv).toBeTruthy();
					expect(Array.isArray(e.errors)).toBeTruthy();
					expect(e.errors.length).toBe(1);

					const error = e.errors[0];
					expect(error.instancePath).toBe(
						'/ubl:Invoice/cac:InvoiceLine/0/cbc:ID',
					);
					expect(error.schemaPath).toBe(
						'#/properties/ubl%3AInvoice/properties/cac%3AInvoiceLine/items/properties/cbc%3AID',
					);
					expect(error.keyword).toBe('type');
					expect(error.params).toEqual({ type: 'string' });
					expect(error.message).toBe(
						"reference '=:Lines.A1' resolves to null: cannot find section 'Lines' in sheet 'Invoice'",
					);
				}

				mockValidateMapping.mockRestore();
			});

			it('should throw an exception if a non-existing sheet is referenced as section', async () => {
				const localMapping = structuredClone(mapping);
				localMapping['ubl:Invoice']['cac:InvoiceLine']['section'] =
					'Infoice:Line';
				const mockValidateMapping = jest
					.spyOn(service as any, 'validateMapping')
					.mockReturnValue(localMapping);
				const buf: Uint8Array = [] as unknown as Uint8Array;

				jest.spyOn(XLSX, 'read').mockReturnValueOnce(workbook);

				try {
					service.transform(buf, 'UBL', {} as Mapping);
					throw new Error('no exception thrown');
				} catch (e) {
					expect(e).toBeDefined();
					expect(e.validation).toBeTruthy();
					expect(e.ajv).toBeTruthy();
					expect(Array.isArray(e.errors)).toBeTruthy();
					expect(e.errors.length).toBe(1);

					const error = e.errors[0];
					expect(error.instancePath).toBe(
						'/ubl:Invoice/cac:InvoiceLine/section',
					);
					expect(error.schemaPath).toBe(
						'#/properties/ubl%3AInvoice/properties/cac%3AInvoiceLine/properties/section',
					);
					expect(error.keyword).toBe('type');
					expect(error.params).toEqual({ type: 'string' });
					expect(error.message).toBe(
						"section reference 'Infoice:Line' resolves to null: no section column for sheet 'Infoice'",
					);
				}

				mockValidateMapping.mockRestore();
			});

			it('should throw an exception if a non-existing section is referenced as section', async () => {
				const localMapping = structuredClone(mapping);
				localMapping['ubl:Invoice']['cac:InvoiceLine']['section'] =
					'Invoice:Lime';
				const mockValidateMapping = jest
					.spyOn(service as any, 'validateMapping')
					.mockReturnValue(localMapping);
				const buf: Uint8Array = [] as unknown as Uint8Array;

				jest.spyOn(XLSX, 'read').mockReturnValueOnce(workbook);

				try {
					service.transform(buf, 'UBL', {} as Mapping);
					throw new Error('no exception thrown');
				} catch (e) {
					expect(e).toBeDefined();
					expect(e.validation).toBeTruthy();
					expect(e.ajv).toBeTruthy();
					expect(Array.isArray(e.errors)).toBeTruthy();
					expect(e.errors.length).toBe(1);

					const error = e.errors[0];
					expect(error.instancePath).toBe(
						'/ubl:Invoice/cac:InvoiceLine/section',
					);
					expect(error.schemaPath).toBe(
						'#/properties/ubl%3AInvoice/properties/cac%3AInvoiceLine/properties/section',
					);
					expect(error.keyword).toBe('type');
					expect(error.params).toEqual({ type: 'string' });
					expect(error.message).toBe(
						"section reference 'Invoice:Lime' resolves to null: no section 'Lime' in sheet 'Invoice'",
					);
				}

				mockValidateMapping.mockRestore();
			});

			it('should throw an exception if multiple unbound sections are encountered', async () => {
				const mockValidateMapping = jest
					.spyOn(service as any, 'validateMapping')
					.mockReturnValue(mapping);
				const buf: Uint8Array = [] as unknown as Uint8Array;

				const localWorkbook = structuredClone(workbook);
				localWorkbook.Sheets.Invoice.K28 = { t: 's', v: 'SubTotal' };

				jest.spyOn(XLSX, 'read').mockReturnValueOnce(localWorkbook);

				try {
					service.transform(buf, 'UBL', {} as Mapping);
					throw new Error('no exception thrown');
				} catch (e) {
					expect(e).toBeDefined();
					expect(e.validation).toBeTruthy();
					expect(e.ajv).toBeTruthy();
					expect(Array.isArray(e.errors)).toBeTruthy();
					expect(e.errors.length).toBe(1);

					const error = e.errors[0];
					expect(error.instancePath).toBe(
						'/ubl:Invoice/cac:LegalMonetaryTotal/cbc:LineExtensionAmount',
					);
					expect(error.schemaPath).toBe(
						'#/properties/ubl%3AInvoice/properties/cac%3ALegalMonetaryTotal/properties/cbc%3ALineExtensionAmount',
					);
					expect(error.keyword).toBe('type');
					expect(error.params).toEqual({ type: 'string' });
					expect(error.message).toBe(
						"reference '=:SubTotal.F1' resolves to null: multiple unbound sections 'SubTotal' in sheet 'Invoice'",
					);
				}

				mockValidateMapping.mockRestore();
			});

			// Test the parseMapping() method.
			it('should validate mappings and fill defaults', () => {
				const formatter: EInvoiceFormat = {
					fillMappingDefaults: jest.fn(),
				} as unknown as EInvoiceFormat;
				const formatFactorySpy = jest
					.spyOn(FormatFactoryService.prototype, 'createFormatService')
					.mockReturnValue(formatter);
				const validateSpy = jest
					.spyOn(ValidationService.prototype, 'validate')
					.mockReturnValue({ 'ubl:Invoice': {} });

				const buf: Uint8Array = [] as unknown as Uint8Array;
				jest.spyOn(XLSX, 'read').mockReturnValueOnce(workbook);

				const mockFillSectionRanges = jest
					.spyOn(service as any, 'fillSectionRanges')
					.mockImplementationOnce(() => {});

				const invoice = service.transform(buf, 'ZIRKUSFeRD', {} as Mapping);

				expect(invoice).toEqual({ 'ubl:Invoice': {} });
				expect(validateSpy).toHaveBeenCalledTimes(1);
				expect(validateSpy).toHaveBeenCalledWith(
					'mapping data',
					expect.anything(),
					{},
				);

				expect(formatFactorySpy).toHaveBeenCalledTimes(1);
				expect(formatFactorySpy).toHaveBeenCalledWith('ZIRKUSFeRD', logger);
				expect(formatter.fillMappingDefaults).toHaveBeenCalledTimes(1);
				expect(formatter.fillMappingDefaults).toHaveBeenCalledWith({
					'ubl:Invoice': {},
				});

				formatFactorySpy.mockRestore();
				validateSpy.mockRestore();
				mockFillSectionRanges.mockRestore();
			});

			describe('should transform invoice data', () => {
				let invoice: Invoice;

				beforeAll(async () => {
					const mockValidateMapping = jest
						.spyOn(service as any, 'validateMapping')
						.mockReturnValue(mapping);
					const buf: Uint8Array = [] as unknown as Uint8Array;

					jest.spyOn(XLSX, 'read').mockReturnValueOnce(workbook);

					invoice = service.transform(buf, 'UBL', {} as Mapping);

					mockValidateMapping.mockRestore();
				});

				it('should return an invoice object', () => {
					expect(invoice).toBeDefined();
					expect(invoice['ubl:Invoice']).toBeDefined();
				});

				it('should map a literal value', () => {
					expect(invoice['ubl:Invoice']['cbc:ID']).toBe('1234567890');
					expect(typeof invoice['ubl:Invoice']['cbc:ID']).toBe('string');
				});

				it('should map a literal date', () => {
					expect(invoice['ubl:Invoice']['cbc:IssueDate']).toBe('2024-10-21');
				});

				it('should map a date', () => {
					expect(invoice['ubl:Invoice']['cbc:DueDate']).toBe('2024-10-28');
				});

				it('should map a string value', () => {
					const wanted = workbook.Sheets.Invoice.B1.v;
					expect(invoice['ubl:Invoice']['cbc:CustomizationID']).toBe(wanted);
				});

				it('should map a quoted literal value', () => {
					const wanted = '= not equals eagles';
					expect(invoice['ubl:Invoice']['cbc:Note']).toBe(wanted);
				});

				it('should map a nested object property', () => {
					const wanted = '0815';
					expect(invoice['ubl:Invoice']['cac:OrderReference']?.['cbc:ID']).toBe(
						wanted,
					);
				});

				it('should map an array of invoice lines', () => {
					const target = invoice['ubl:Invoice']['cac:InvoiceLine'];
					expect(target).toBeDefined();
					expect(Array.isArray(target)).toBe(true);
					expect(target.length).toBe(3);
					expect(target[0]['cbc:ID']).toBe('1');
					expect(target[1]['cbc:ID']).toBe('2');
					expect(target[2]['cbc:ID']).toBe('3');
				});

				it('should map allowances charges of invoice line #1', () => {
					const target =
						invoice['ubl:Invoice']['cac:InvoiceLine'][0]['cac:AllowanceCharge'];
					expect(target).toBeDefined();
					if (typeof target !== 'undefined') {
						expect(Array.isArray(target)).toBe(true);
						expect(target.length).toBe(1);
						expect(target[0]['cbc:Amount']).toBe('23.04');
						expect(target[0]['cbc:Amount@currencyID']).toBe('EUR');
					}
				});

				it('should map allowances charges of invoice line #2', () => {
					const target =
						invoice['ubl:Invoice']['cac:InvoiceLine'][1]['cac:AllowanceCharge'];
					expect(target).toBeDefined();
					if (typeof target !== 'undefined') {
						expect(Array.isArray(target)).toBe(true);
						expect(target.length).toBe(0);
					}
				});

				it('should map allowances charges of invoice line #3', () => {
					const target =
						invoice['ubl:Invoice']['cac:InvoiceLine'][2]['cac:AllowanceCharge'];
					expect(target).toBeDefined();
					if (typeof target !== 'undefined') {
						expect(Array.isArray(target)).toBe(true);
						expect(target.length).toBe(2);
						expect(target[0]['cbc:Amount']).toBe('13.03');
						expect(target[0]['cbc:Amount@currencyID']).toBe('EUR');
						expect(target[1]['cbc:Amount']).toBe('42.00');
						expect(target[1]['cbc:Amount@currencyID']).toBe('EUR');
					}
				});

				it('should map non-array sections', () => {
					const target =
						invoice['ubl:Invoice']['cac:LegalMonetaryTotal'][
							'cbc:LineExtensionAmount'
						];
					expect(target).toBe('12345.67');
				});

				it('should map arrays without sections', () => {
					const target = invoice['ubl:Invoice']['cac:PaymentMeans'];
					expect(Array.isArray(target)).toBe(true);
					expect(target?.[0]['cbc:PaymentMeansCode']).toBe('30');
				});
			});
		});
	});

	describe('migration', () => {
		it('should migrate mapping v1 to v3', () => {
			const mockValidateMapping = jest
			.spyOn(service as any, 'validateMapping')
			.mockReturnValue(mapping);

			const migrated = service.migrate(mapping);
			expect(migrated).toBeDefined();

			mockValidateMapping.mockRestore();
		});
	});
});
