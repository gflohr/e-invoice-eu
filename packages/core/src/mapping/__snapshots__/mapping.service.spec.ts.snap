// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MappingService migration should migrate mapping v1 to v3 1`] = `
{
  "meta": {
    "sectionColumn": {
      "Invoice": "K",
      "Unused": "L",
    },
    "version": 3,
  },
  "ubl:Invoice": {
    "cac:InvoiceLine": {
      "cac:AllowanceCharge": {
        "cbc:Amount": "=SECTIONVALUE("B1", "ACLine")",
        "cbc:Amount@currencyID": "=SECTIONVALUE("B1", "Line")",
        "section": ":ACLine",
      },
      "cbc:ID": "=SECTIONVALUE("A1", "Line")",
      "section": ":Line",
    },
    "cac:LegalMonetaryTotal": {
      "cbc:LineExtensionAmount": "=SECTIONVALUE("F1", "SubTotal")",
    },
    "cac:OrderReference": {
      "cbc:ID": "=A3",
    },
    "cac:PaymentMeans": {
      "cbc:PaymentMeansCode": "30",
    },
    "cac:TaxTotal": {
      "cbc:TaxAmount@currencyID": "EUR",
    },
    "cbc:CustomizationID": "=B1",
    "cbc:DueDate": "='Invoice'.A2",
    "cbc:ID": "1234567890",
    "cbc:IssueDate": "=A1",
    "cbc:Note": "'= not equals eagles",
  },
}
`;

exports[`MappingService v3 should parse mappings and map data should generate the correct invoice 1`] = `
{
  "ubl:Invoice": {
    "cac:InvoiceLine": [
      {
        "cac:AllowanceCharge": [
          {
            "cbc:Amount": "23.04",
            "cbc:Amount@currencyID": "EUR",
          },
        ],
        "cbc:ID": "1",
      },
      {
        "cac:AllowanceCharge": [],
        "cbc:ID": "2",
      },
      {
        "cac:AllowanceCharge": [
          {
            "cbc:Amount": "13.03",
            "cbc:Amount@currencyID": "EUR",
          },
          {
            "cbc:Amount": "42.00",
            "cbc:Amount@currencyID": "EUR",
          },
        ],
        "cbc:ID": "3",
      },
    ],
    "cac:LegalMonetaryTotal": {
      "cbc:LineExtensionAmount": "12345.67",
    },
    "cac:OrderReference": {
      "cbc:ID": "0815",
    },
    "cac:PaymentMeans": [
      {
        "cbc:PaymentMeansCode": "30",
      },
    ],
    "cac:TaxTotal": [
      {},
    ],
    "cbc:CustomizationID": "urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0",
    "cbc:DueDate": "2024-10-28",
    "cbc:ID": "1234567890",
    "cbc:IssueDate": "2024-10-21",
    "cbc:Note": "= not equals eagles",
  },
}
`;
